include/swift/Basic/Assertions.h

```diff
@@ -0,0 +1,181 @@
+// This should be used for most assertions, which is why it
+// deserves the short name.  In particular, for simple checks
+// (e.g., validating that something is non-null), this is just as
+// fast as a disabled `CONDITIONAL_ASSERT`, so there's no point in
+// using the conditional form.
+//
+// You can globally replace `assert` with `ASSERT` in a piece of code
+// to have all your assertions enabled in all builds.  If you do this,
+// please do a little profiling first, just in case you have some checks
+// that are more expensive than you think.  You can switch those to
+// `CONDITIONAL_ASSERT` or `DEBUG_ASSERT` as needed.
+
+#define ASSERT(expr) \
+  do { \
+    if (!(expr)) { \
+      ASSERT_failure(#expr, __FILE__, __LINE__, __func__); \
+    } \
+  } while (0)
+
+// Function that reports the actual failure when it occurs.
+void ASSERT_failure(const char *expr, const char *file, int line, const char *func);
+
+// ================================ Conditional Asserts ================================
+
+// `CONDITIONAL_ASSERT(expr)`:
+// * is always compiled into the executable, but
+// * only checks the condition if a runtime flag is defined.
+// That runtime flag is disabled by default in release builds
+// but can be enabled with the command-line flag `-compiler-assertions`
+//
+// Use this for asserts that are comparatively expensive to check.
+//
+// You can globally change `assert` to `CONDITIONAL_ASSERT` to make all your
+// assertions _optionally_ available in release builds.  Anyone can then add
+// `-compiler-assertions` to their build flags to get more information about a
+// compiler problem.  Before you check it in, do a little checking for
+// assertions that might be checked huge numbers of times (e.g., invariants
+// for inner loops or core utilities); those may need to become `DEBUG_ASSERT`
+// or else refactored to be checked more selectively.
+//
+// Over time, plan to change most of the resulting `CONDITIONAL_ASSERT` into
+// plain `ASSERT` to enable them by default.
+
+#define CONDITIONAL_ASSERT(expr) \
+  do { \
+    if (CONDITIONAL_ASSERT_enabled()) { \
+      ASSERT(expr); \
+    } \
+  } while (0)
+
+// Use `CONDITIONAL_ASSERT_enabled()` to guard complex, expensive code that
+// should only run when assertions are enabled.  This is exactly the
+// same check that's used to enable `CONDITIONAL_ASSERT()` at runtime.
+// This is not often used -- if you are just setting a flag or updating
+// a counter, it's likely cheaper just to do it than to test whether or not
+// to do it. Only use this for relatively complex operations.
+//
+// if (CONDITIONAL_ASSERT_enabled()) {
+//    ... stuff ...
+// }
+
+// Declare a callable function version of this runtime test.
+int CONDITIONAL_ASSERT_enabled();
+
+// Define a macro version of this test
+extern int CONDITIONAL_ASSERT_Global_enable_flag;
+#define CONDITIONAL_ASSERT_enabled() \
+  (CONDITIONAL_ASSERT_Global_enable_flag != 0)
```

Comment:
Wrapping this in `SWIFT_UNLIKELY` might help the no-asserts case slightly.

Reply:
Done.